### PR TITLE
Second Sprint Draft PR

### PR DESCRIPTION
**Pull request by samcarley**
This PR consists of our project write-up (implemented within CodeChat), our first test case (via cargo test), and a function that is similar to "save endpoint" but returns the file contents in a string. This is so we may test the save functionality within our testing functions.

### CODE REVIEW

File: server/src/webserver.rs [hunk 1]
```diff
@@ -255,6 +255,140 @@ async fn save_source(
     save_source_response(true, "")
 }
 
+    //#[put("/fs/{path:.*}")]
```
Comment (bjones1) [inline server/src/webserver.rs:258]: Delete this.

File: server/src/webserver.rs [hunk 2]
```diff
@@ -255,6 +255,140 @@ async fn save_source(
...
+    // <h3>Testing input to save file</h3>
+    // <p>This function is only used for testing; it does not actively save
```
Comment (bjones1) [inline server/src/webserver.rs:260]: Change/fix -- this should be called by the `save_source` function, so the production code is tested.

File: server/src/webserver.rs [hunk 3]
```diff
@@ -844,3 +987,33 @@ pub async fn main() -> std::io::Result<()> {
...
+#[cfg(test)]
+
+
+// <p>Provide a way to create a CodeDocBlock</p>
+
+
+    
+// <h3>Save Endpoint Testing</h3>
+
+#[test]
```
Comment (bjones1) [inline server/src/webserver.rs:1010]: Put all tests in side a `mod test { ... }` like the lexer's tests.

**Review by bjones1 — CHANGES_REQUESTED** [review]

File: server/src/webserver.rs (hunk 1 above)
Comment (samcarley) [reply server/src/webserver.rs]: Done!

File: server/src/webserver.rs (hunk 2 above)
Comment (samcarley) [reply server/src/webserver.rs]: Done!

File: server/src/webserver.rs (hunk 3 above)
Comment (samcarley) [reply server/src/webserver.rs]: Added!

**Comment by samcarley** [comment]
Implemented comments/requested changes.

File: server/src/webserver.rs [hunk 4]
```diff
@@ -110,13 +110,51 @@ async fn save_source(
...
+    language_lexers_compiled: web::Data<LanguageLexersCompiled<'_>>,
+) -> (String, impl Responder) {
     // <p>Given the mode, find the lexer.</p>
     let lexer = match language_lexers_compiled
         .map_mode_to_lexer
         .get(client_source_file.metadata.mode.as_str())
     {
         Some(v) => v,
-        None => return save_source_response(false, "Invalid mode"),
+        None => return (String::new(), save_source_response(false, "Invalid mode")),
```
Comment (bjones1) [inline server/src/webserver.rs:157]: OK for this PR, but we need to think about error handling here. I'm probably recommend returning a [result](https://doc.rust-lang.org/std/result/), then unwrapping that in the calling function.

File: server/src/webserver.rs [hunk 5]
```diff
@@ -144,10 +182,10 @@ async fn save_source(
                         } else if let Some(bc) = block_comment {
                             bc.opening.to_string()
                         } else {
-                            return save_source_response(
+                            return (String::new(), save_source_response(
```
Comment (bjones1) [inline server/src/webserver.rs:185]: Same comment here -- all error handling should instead be handled as a result.

File: server/src/webserver.rs [hunk 6]
```diff
@@ -844,3 +871,34 @@ pub async fn main() -> std::io::Result<()> {
...
+// <p>&nbsp;</p>
+#[cfg(test)]
+
+
+// <p>Provide a way to create a CodeDocBlock</p>
+
+
+    
+// <h3>Save Endpoint Testing</h3>
+mod tests {
```
Comment (bjones1) [inline server/src/webserver.rs:893]: A good start. For the next PR, run `cargo fmt` and `cargo clippy` to help clean up the source.

**Pull request closed without merging**